package com.pia.tmf.v4.common.util;

import static com.pia.client.common.util.WebClientUtil.retry;
import static com.pia.tmf.v4.common.util.TmfClientCommonHeaderUtil.getHeadersFirstValue;
import static org.springframework.util.StringUtils.hasText;

import com.github.fge.jsonpatch.JsonPatch;
import com.jayway.jsonpath.JsonPath;
import com.pia.client.common.model.BaseClientProperties;
import com.pia.client.common.service.api.TokenService;
import com.pia.client.common.util.WebClientUtil;
import com.pia.commons.util.JacksonUtil;
import com.pia.tmf.v4.common.config.TmfClientConfigurations;
import com.pia.tmf.v4.common.config.TmfClientConfigurations.TmfClientConfig;
import com.pia.tmf.v4.common.exception.TmfClientException;
import com.pia.tmf.v4.common.model.Error;
import com.pia.tmf.v4.common.model.JsonFilter;
import com.pia.tmf.v4.common.model.OffsetPage;
import com.pia.tmf.v4.common.model.RetrievalContext;
import com.pia.tmf.v4.common.model.Scope;
import com.pia.tmf.v4.common.model.TmfClientCommonsConstants;
import com.pia.tmf.v4.common.model.TmfOffsetRequest;
import com.pia.tmf.v4.common.model.TmfPage;
import java.net.URI;
import java.time.Duration;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.function.Consumer;
import java.util.function.Function;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.client.ClientResponse;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.util.UriComponentsBuilder;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * Utility class for common operations used in TMF (Telecom Management Forum) client applications.
 * This class provides methods for building URIs, making HTTP requests, handling pagination,
 * registering and unregistering event listeners, and other common functionalities.
 *
 * <p>Author: Isra
 *
 * <p>Author: Yusuf BOZKURT
 */
@Slf4j
public final class TmfClientCommonUtil {

  public static final String FILTER = "filter";
  public static final String FIELDS = "fields";

  /** Private constructor to prevent instantiation of this utility class. */
  private TmfClientCommonUtil() {}

  /**
   * Retrieves an authentication token for the specified client path from the given TokenService. If
   * the client path has a scope defined, the token is retrieved with that scope; otherwise, a token
   * is retrieved without a specific scope.
   *
   * @param tokenService The TokenService used to retrieve the token.
   * @return A Mono emitting the authentication token as a String.
   */
  public static Mono<String> getToken(
      Scope scopeKey, TmfClientConfig config, TokenService tokenService) {
    String scope = config.getScopes().get(scopeKey);
    return hasText(scope) ? tokenService.getToken(scope) : tokenService.getToken();
  }

  /**
   * Builds a URI based on the provided TMF client configuration and client path.
   *
   * @param config The TMF client configuration containing base URL and context path.
   * @return The constructed URI.
   */
  public static URI buildUri(TmfClientConfigurations.TmfClientConfig config) {

    return UriComponentsBuilder.fromUriString(config.getBaseUrl())
        .path(config.getContextPath())
        .path(config.getEndpoint())
        .build()
        .toUri();
  }

  /**
   * Builds a URI with an ID parameter based on the provided TMF client configuration, client path,
   * and ID value.
   *
   * @param config The TMF client configuration containing base URL and context path.
   * @param id The ID value to be included in the URI.
   * @return The constructed URI.
   */
  public static URI buildUriWithId(TmfClientConfigurations.TmfClientConfig config, String id) {
    Objects.requireNonNull(id, TmfClientCommonsConstants.ERROR_MSG_ID_RESPONSE);
    return UriComponentsBuilder.fromUriString(config.getBaseUrl())
        .path(config.getContextPath())
        .path(config.getEndpoint())
        .path("/{id}")
        .build(id);
  }

  /**
   * Builds a URI with an ID parameter and additional filter parameters based on the provided TMF
   * client configuration, client path, ID value, and filter.
   *
   * @param config The TMF client configuration containing base URL and context path.
   * @param id The ID value to be included in the URI.
   * @param retrievalContext The filter object containing filter parameters.
   * @return The constructed URI.
   */
  public static URI buildUriWithId(
      TmfClientConfigurations.TmfClientConfig config,
      String id,
      RetrievalContext retrievalContext) {
    Objects.requireNonNull(id, TmfClientCommonsConstants.ERROR_MSG_ID_RESPONSE);
    UriComponentsBuilder uriComponentsBuilder =
        UriComponentsBuilder.fromUriString(config.getBaseUrl())
            .path(config.getContextPath())
            .path(config.getEndpoint())
            .path("/{id}");

    if (retrievalContext.getJsonFilterType() == JsonFilter.TYPE.SERVER) {
      uriComponentsBuilder.queryParam(FILTER, retrievalContext.getJsonFilterQuery());
    }

    if (retrievalContext.getFields() != null && !retrievalContext.getFields().isEmpty()) {
      uriComponentsBuilder.queryParam(FIELDS, prepareFieldsQuery(retrievalContext.getFields()));
    }
    return uriComponentsBuilder.build(id);
  }

  /**
   * Builds a URI with query parameters based on the provided TMF client configuration, client path,
   * and query parameters.
   *
   * @param config The TMF client configuration containing base URL and context path.
   * @param queryParams The query parameters to be included in the URI.
   * @return The constructed URI.
   */
  public static URI buildUri(
      TmfClientConfigurations.TmfClientConfig config, MultiValueMap<String, String> queryParams) {
    return UriComponentsBuilder.fromUriString(config.getBaseUrl())
        .path(config.getContextPath())
        .path(config.getEndpoint())
        .queryParams(queryParams)
        .build()
        .toUri();
  }

  private static URI updateUri(URI uri, RetrievalContext retrievalContext) {
    UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromUri(uri);
    if (retrievalContext.getJsonFilterType() == JsonFilter.TYPE.SERVER) {
      uriComponentsBuilder.queryParam(FILTER, retrievalContext.getJsonFilterQuery());
    }
    if (retrievalContext.getFields() != null && !retrievalContext.getFields().isEmpty()) {
      uriComponentsBuilder.queryParam(FIELDS, prepareFieldsQuery(retrievalContext.getFields()));
    }
    return uriComponentsBuilder.build().toUri();
  }

  private static URI updateUri(URI uri, Pageable pageable) {
    UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromUri(uri);
    uriComponentsBuilder.queryParam("offset", pageable.getOffset());
    uriComponentsBuilder.queryParam("limit", pageable.getPageSize());
    if (!pageable.getSort().isEmpty()) {
      String sortQuery = sortStringQuery(pageable.getSort());
      uriComponentsBuilder.queryParam("sort", sortQuery);
    }

    if (pageable instanceof TmfOffsetRequest tmfOffsetRequest) {
      if (tmfOffsetRequest.getJsonFilterTYpe() == JsonFilter.TYPE.SERVER) {
        uriComponentsBuilder.queryParam(FILTER, tmfOffsetRequest.getJsonFilterQuery());
      }
      if (tmfOffsetRequest.getFields() != null && !tmfOffsetRequest.getFields().isEmpty()) {
        uriComponentsBuilder.queryParam(FIELDS, prepareFieldsQuery(tmfOffsetRequest.getFields()));
      }
    }
    return uriComponentsBuilder.build().toUri();
  }

  private static String sortStringQuery(Sort sort) {
    return sort.stream()
        .map(
            order -> (order.getDirection() == Sort.Direction.DESC ? "-" : "") + order.getProperty())
        .reduce((a, b) -> a + "," + b)
        .orElse("");
  }

  private static String prepareFieldsQuery(Set<String> fields) {
    return fields.stream().reduce((a, b) -> a + "," + b).orElse("");
  }

  /**
   * Executes a POST request to the specified URI with the provided body and access token, handling
   * errors and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the POST request is made.
   * @param body The request body.
   * @param headersConsumer The authentication token used for authorization.
   * @param t The class type of the expected response.
   * @param errorhandler Function to handle errors.
   * @param properties The properties governing the behavior of the request.
   * @return A Mono emitting the response object.
   */
  public static <T> Mono<T> postRequest(
      WebClient webClient,
      URI uri,
      Object body,
      Consumer<HttpHeaders> headersConsumer,
      Class<T> t,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      BaseClientProperties properties) {
    Objects.requireNonNull(
        body, TmfClientCommonsConstants.ERROR_MSG_EMPTY_RESPONSE.formatted(t.getSimpleName()));
    validateHeadersConsumer(headersConsumer);
    return webClient
        .post()
        .uri(uri)
        .headers(headersConsumer)
        .bodyValue(body)
        .retrieve()
        .onStatus(HttpStatusCode::isError, errorhandler)
        .bodyToMono(t)
        .retryWhen(
            retry(properties.getNumRetries(), Duration.ofMillis(properties.getRetryWaitMillis())));
  }

  /**
   * Executes a PATCH request to the specified URI with the provided JSON patch and access token,
   * handling errors and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the PATCH request is made.
   * @param jsonPatch The JSON patch to apply.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param t The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param <T> The type of the expected response.
   * @return A Mono emitting the response object.
   * @throws NullPointerException if any of the required parameters is null.
   */
  public static <T> Mono<T> patchRequest(
      WebClient webClient,
      URI uri,
      JsonPatch jsonPatch,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties) {
    Objects.requireNonNull(
        jsonPatch, TmfClientCommonsConstants.ERROR_MSG_EMPTY_RESPONSE.formatted(t.getSimpleName()));
    validateHeadersConsumer(headersConsumer);
    return webClient
        .patch()
        .uri(uri)
        .headers(headersConsumer)
        .contentType(MediaType.valueOf(TmfClientCommonsConstants.MEDIA_TYPE_JSON_PATCH))
        .bodyValue(jsonPatch)
        .retrieve()
        .onStatus(HttpStatusCode::isError, errorhandler)
        .bodyToMono(t)
        .retryWhen(
            WebClientUtil.retry(
                properties.getNumRetries(), Duration.ofMillis(properties.getRetryWaitMillis())));
  }

  /**
   * Executes a PATCH request to the specified URI with the provided body and access token, handling
   * errors and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the PATCH request is made.
   * @param body The request body.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param t The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param <T> The type of the expected response.
   * @return A Mono emitting the response object.
   * @throws NullPointerException if any of the required parameters is null.
   */
  public static <T> Mono<T> patchRequest(
      WebClient webClient,
      URI uri,
      Object body,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties) {
    Objects.requireNonNull(
        body, TmfClientCommonsConstants.ERROR_MSG_EMPTY_RESPONSE.formatted(t.getSimpleName()));
    validateHeadersConsumer(headersConsumer);
    return webClient
        .patch()
        .uri(uri)
        .headers(headersConsumer)
        .contentType(MediaType.valueOf(TmfClientCommonsConstants.MEDIA_TYPE_MERGE_PATCH))
        .bodyValue(body)
        .retrieve()
        .onStatus(HttpStatusCode::isError, errorhandler)
        .bodyToMono(t)
        .retryWhen(
            WebClientUtil.retry(
                properties.getNumRetries(), Duration.ofMillis(properties.getRetryWaitMillis())));
  }

  /**
   * Executes a GET request to the specified URI with the provided access token and filter, handling
   * errors and retries as per the provided properties. If a client-side filter is provided, it
   * processes the response accordingly.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the GET request is made.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param t The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param retrievalContext The filter to apply to the response.
   * @param <T> The type of the expected response.
   * @return A Mono emitting the response object.
   * @throws NullPointerException if any of the required parameters is null.
   */
  public static <T> Mono<T> getRequest(
      WebClient webClient,
      URI uri,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties,
      RetrievalContext retrievalContext) {

    URI updateUri = updateUri(uri, retrievalContext);
    if (retrievalContext.getJsonFilterType() == JsonFilter.TYPE.CLIENT) {
      return getRequestWithClientFilter(
          webClient, updateUri, headersConsumer, errorhandler, t, properties, retrievalContext);
    }
    return getRequest(webClient, updateUri, headersConsumer, errorhandler, t, properties);
  }

  public static <T> Mono<T> getRequest(
      WebClient webClient,
      URI uri,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties) {
    validateHeadersConsumer(headersConsumer);
    return webClient
        .get()
        .uri(uri)
        .headers(headersConsumer)
        .retrieve()
        .onStatus(HttpStatusCode::isError, errorhandler)
        .bodyToMono(t)
        .retryWhen(
            WebClientUtil.retry(
                properties.getNumRetries(), Duration.ofMillis(properties.getRetryWaitMillis())));
  }

  private static <T> Mono<T> getRequestWithClientFilter(
      WebClient webClient,
      URI uri,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties,
      RetrievalContext retrievalContext) {
    return getRequest(webClient, uri, headersConsumer, errorhandler, Object.class, properties)
        .map(object -> JsonPath.read(object, retrievalContext.getJsonFilterQuery()));
  }

  /**
   * Executes a GET request to the specified URI with the provided access token, handling errors and
   * retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the GET request is made.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param t The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   * @throws NullPointerException if any of the required parameters is null.
   */
  public static <T> Flux<T> getAllRequest(
      WebClient webClient,
      URI uri,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties) {
    return retrieveSinglePage(webClient, uri, headersConsumer, errorhandler, t, properties);
  }

  /**
   * Retrieves all pages of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   */
  public static <T> Flux<T> retrieveAllPages(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties) {
    return retrieveAllPages(
        webClient, url, headersConsumer, errorhandler, type, properties, TmfOffsetRequest.of());
  }

  /**
   * Retrieves all pages of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties and applying the given page query.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param pageQuery The page query specifying pagination parameters.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   */
  public static <T> Flux<T> retrieveAllPages(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties,
      Pageable pageQuery) {

    if (pageQuery instanceof TmfOffsetRequest tmfOffsetRequest
        && (tmfOffsetRequest.getJsonFilterTYpe() == JsonFilter.TYPE.CLIENT)) {
        return retrieveAllPagesWithClientFilter(
            webClient, url, headersConsumer, errorhandler, type, properties, tmfOffsetRequest);

    }

    return retrieveAllPagesWithServerFilter(
        webClient, url, headersConsumer, errorhandler, type, properties, pageQuery);
  }

  /**
   * Retrieves all pages of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties, and applying client-side filtering.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param tmfOffsetRequest The offset request containing client-side filtering information.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   */
  private static <T> Flux<T> retrieveAllPagesWithClientFilter(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties,
      TmfOffsetRequest tmfOffsetRequest) {

    return recursiveRetrieve(
            webClient,
            url,
            headersConsumer,
            errorhandler,
            Object.class,
            properties,
            tmfOffsetRequest)
        .collectList()
        .flatMapMany(
            objects -> {
              List<T> result = JsonPath.read(objects, tmfOffsetRequest.getJsonFilter().getQuery());
              return Flux.fromIterable(
                  result.stream()
                      .map(o -> JacksonUtil.jsonToObject(JacksonUtil.objectToJson(o), type))
                      .toList());
            });
  }

  /**
   * Retrieves all pages of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties, and applying server-side filtering.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param pageQuery The page query specifying pagination parameters.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   */
  private static <T> Flux<T> retrieveAllPagesWithServerFilter(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties,
      Pageable pageQuery) {
    return recursiveRetrieve(
        webClient, url, headersConsumer, errorhandler, type, properties, pageQuery);
  }

  /**
   * Retrieves a single page of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   */
  public static <T> Flux<T> retrieveSinglePage(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties) {
    return retrieveSinglePage(
        webClient, url, headersConsumer, errorhandler, type, properties, TmfOffsetRequest.of(0));
  }

  /**
   * Retrieves a single page of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties, and applying the given page query.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param pageQuery The page query specifying pagination parameters.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   */
  public static <T> Flux<T> retrieveSinglePage(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties,
      Pageable pageQuery) {
    return retrieveSinglePageWithPageResponse(
            webClient, url, headersConsumer, errorhandler, type, properties, pageQuery)
        .flatMapMany(TmfPage::getContent);
  }

  /**
   * Retrieves a single page of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param <T> The type of the expected response.
   * @return A Mono emitting the response page.
   */
  public static <T> Mono<TmfPage<Flux<T>>> retrieveSinglePageWithPageResponse(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties) {
    Pageable pageable = TmfOffsetRequest.of(0);
    return retrieveSinglePageWithPageResponse(
        webClient, url, headersConsumer, errorhandler, type, properties, pageable);
  }

  /**
   * Retrieves a single page of data from the specified URI with the provided access token, handling
   * errors and retries as per the provided properties, and applying the given offset request.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param offsetRequest The offset request specifying pagination parameters.
   * @param <T> The type of the expected response.
   * @return A Mono emitting the response page.
   */
  public static <T> Mono<TmfPage<Flux<T>>> retrieveSinglePageWithPageResponse(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties,
      Pageable offsetRequest) {
    return getAllRequestWithResponseEntity(
            webClient, url, headersConsumer, errorhandler, type, properties, offsetRequest)
        .map(responseEntity -> preparePagedResponse(responseEntity, offsetRequest));
  }

  /**
   * Retrieves the response entity containing a Flux of data from the specified URI with the
   * provided access token, handling errors and retries as per the provided properties, and applying
   * the given page query.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the GET request is made.
   * @param httpHeadersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param t The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param pageQuery The page query specifying pagination parameters.
   * @param <T> The type of the expected response.
   * @return A Mono emitting the response entity. with header, pageable
   */
  public static <T> Mono<ResponseEntity<Flux<T>>> getAllRequestWithResponseEntity(
      WebClient webClient,
      URI uri,
      Consumer<HttpHeaders> httpHeadersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties,
      Pageable pageQuery) {
    URI updatedUri = updateUri(uri, pageQuery);
    return getAllRequestWithResponseEntity(
        webClient, updatedUri, httpHeadersConsumer, errorhandler, t, properties);
  }

  /**
   * Retrieves the response entity containing a Flux of data from the specified URI with the
   * provided access token, handling errors and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the GET request is made.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param t The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param <T> The type of the expected response.
   * @return A Mono emitting the response entity.
   */
  public static <T> Mono<ResponseEntity<Flux<T>>> getAllRequestWithResponseEntity(
      WebClient webClient,
      URI uri,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> t,
      BaseClientProperties properties) {
    Objects.requireNonNull(t, TmfClientCommonsConstants.ERROR_MSG_NULL_RETURN_CLASS_TYPE);
    validateHeadersConsumer(headersConsumer);
    return webClient
        .get()
        .uri(uri)
        .headers(headersConsumer)
        .retrieve()
        .onStatus(HttpStatusCode::isError, errorhandler)
        .toEntityFlux(t)
        .retryWhen(
            WebClientUtil.retry(
                properties.getNumRetries(), Duration.ofMillis(properties.getRetryWaitMillis())));
  }

  /**
   * Executes a DELETE request to the specified URI with the provided access token, handling errors
   * and retries as per the provided properties.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param uri The URI to which the DELETE request is made.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param properties The properties governing the behavior of the request.
   * @return A Mono emitting when the operation is complete.
   * @throws NullPointerException if any of the required parameters is null.
   */
  public static Mono<Void> deleteRequest(
      WebClient webClient,
      URI uri,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      BaseClientProperties properties) {
    validateHeadersConsumer(headersConsumer);
    return webClient
        .delete()
        .uri(uri)
        .headers(headersConsumer)
        .retrieve()
        .onStatus(HttpStatusCode::isError, errorhandler)
        .toBodilessEntity()
        .retryWhen(
            WebClientUtil.retry(
                properties.getNumRetries(), Duration.ofMillis(properties.getRetryWaitMillis())))
        .then();
  }

  /**
   * Recursively retrieves all pages of data from the specified URL with the provided access token,
   * handling errors and retries as per the provided properties, and applying the given page query.
   *
   * @param webClient The WebClient instance to use for making the request.
   * @param url The base URL from which data is retrieved.
   * @param headersConsumer The authentication token used for authorization.
   * @param errorhandler Function to handle errors.
   * @param type The class type of the expected response.
   * @param properties The properties governing the behavior of the request.
   * @param pageQuery The page query specifying pagination parameters.
   * @param <T> The type of the expected response.
   * @return A Flux emitting the response objects.
   */
  private static <T> Flux<T> recursiveRetrieve(
      WebClient webClient,
      URI url,
      Consumer<HttpHeaders> headersConsumer,
      Function<ClientResponse, Mono<? extends Throwable>> errorhandler,
      Class<T> type,
      BaseClientProperties properties,
      Pageable pageQuery) {
    return retrieveSinglePageWithPageResponse(
            webClient, url, headersConsumer, errorhandler, type, properties, pageQuery)
        .flatMapMany(
            pagedResponse -> {
              if (pagedResponse.isLast()) {
                return pagedResponse.getContent();
              } else {
                return Flux.concat(
                    pagedResponse.getContent(),
                    recursiveRetrieve(
                        webClient,
                        url,
                        headersConsumer,
                        errorhandler,
                        type,
                        properties,
                        pagedResponse.getNextPageable()));
              }
            });
  }

  /**
   * Prepares a paged response from the provided response entity and page query.
   *
   * @param responseEntity The response entity containing the data.
   * @param tmfPageable The pageable object representing the page parameters.
   * @param <T> The type of the response data.
   * @return A TmfPage containing the paged response.
   */
  private static <T> TmfPage<Flux<T>> preparePagedResponse(
      ResponseEntity<Flux<T>> responseEntity, Pageable tmfPageable) {
    long total = getXTotalCount(responseEntity);
    int serverItemCount = getContentRange(responseEntity);
    return new OffsetPage<>(total, serverItemCount, tmfPageable, responseEntity.getBody());
  }

  private static long getXTotalCount(ResponseEntity<?> responseEntity) {
    var value = getHeadersFirstValue("X-Total-Count", responseEntity);
    return hasText(value) ? Long.parseLong(value) : 0;
  }

  private static int getContentRange(ResponseEntity<?> responseEntity) {
    var value = getHeadersFirstValue("Content-Range", responseEntity);
    return hasText(value) ? getItemCountFromContentRange(value) : 0;
  }

  private static int getItemCountFromContentRange(String contentRange) {
    String[] parts = contentRange.replace("items ", "").split("/");
    String[] rangeParts = parts[0].split("-");
    int rangeStart = Integer.parseInt(rangeParts[0].trim());
    int rangeEnd = Integer.parseInt(rangeParts[1].trim());
    return rangeEnd - rangeStart + 1;
  }

  public static <T> TmfClientException createException(
      HttpStatusCode httpStatusCode, Class<? extends TmfClientException> exception) {
    try {
      return exception.getDeclaredConstructor(HttpStatusCode.class).newInstance(httpStatusCode);
    } catch (Exception e) {
      throw new IllegalArgumentException("Exception class misses required constructor.", e);
    }
  }

  public static <T> TmfClientException createException(
      HttpStatusCode httpStatusCode, Error error, Class<? extends TmfClientException> exception) {
    try {
      return exception
          .getDeclaredConstructor(HttpStatusCode.class, Error.class)
          .newInstance(httpStatusCode, error);
    } catch (Exception e) {
      throw new IllegalArgumentException("Exception class misses required constructor.", e);
    }
  }

  private static void validateHeadersConsumer(Consumer<HttpHeaders> headersConsumer) {
    if (headersConsumer != null) {
      HttpHeaders headers = new HttpHeaders();
      headersConsumer.accept(headers);
      String authorizationHeader = headers.getFirst(HttpHeaders.AUTHORIZATION);

      if (authorizationHeader == null || StringUtils.isBlank(authorizationHeader)) {
        throw new IllegalArgumentException(TmfClientCommonsConstants.ERROR_MSG_EMPTY_AUTH_TOKEN);
      }
    } else {
      throw new IllegalArgumentException(TmfClientCommonsConstants.ERROR_MSG_NULL_HEADERS_CONSUMER);
    }
  }
}
